ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java

```diff
@@ -0,0 +1,1760 @@
+   * @param msg the message to print on failure if no exception is thrown.
+   * @throws Exception
+   */
+  private void ensureArgumentFailure(CompactionDetails details, String msg) throws Exception {
+    compactor = getCompactor(state.log, DISK_IO_SCHEDULER);
+    compactor.initialize(state.index);
+    try {
+      compactor.compact(details);
+      fail(msg);
+    } catch (IllegalArgumentException e) {
+      // expected. Nothing to do.
+    } finally {
+      compactor.close(0);
+    }
+  }
+
+  // expirationTimeEnforcementTest(), deletionTimeEnforcementTest() and related recovery test helpers
+
+  /**
+   * Sets up state where a fixed number of log segments are written and a fraction of the blobs in these log segments
+   * expire at a time returned by the function.
+   * @return a {@link Pair} that contains the time at which the blobs will expire and the list of segments that should
+   * be compacted.
+   * @throws Exception
+   */
+  private Pair<Long, List<String>> setupStateWithExpiredBlobsAtSpecificTime() throws Exception {
+    refreshState(false, true);
+    // we want at least 3 log segments that are outside the journal and with no invalid records
+    long extraSegmentCountRequired = 4;
+    // send +1 to account for half filled segments from curated log-index state.
+    long expiryTimeMs = getInvalidationTime(extraSegmentCountRequired + 1);
+
+    long currentLogSegmentCount = state.index.getLogSegmentCount();
+    writeDataToMeetRequiredSegmentCount(currentLogSegmentCount + extraSegmentCountRequired,
+        Collections.singletonList(expiryTimeMs));
+    List<String> segmentsUnderCompaction = getLogSegments(currentLogSegmentCount, extraSegmentCountRequired - 1);
+    // reload index to make sure journal is on only the latest log segment
+    state.reloadIndex(true, false);
+    return new Pair<>(expiryTimeMs, segmentsUnderCompaction);
+  }
+
+  /**
+   * Sets up state where a fixed number of log segments are written and a fraction of the blobs in these log segments
+   * are considered deleted at a time returned by the function.
+   * @return a {@link Pair} that contains the time at which the blobs are considered deleted and the list of segments
+   * that should be compacted.
+   * @throws Exception
+   */
+  private Pair<Long, List<String>> setupStateWithDeletedBlobsAtSpecificTime() throws Exception {
+    refreshState(false, true);
+    // we want at least 3 log segments that are outside the journal and with no invalid records
+    long extraSegmentCountRequired = 4;
+    // send +1 to account for half filled segments from curated log-index state.
+    long deleteTimeMs = getInvalidationTime(extraSegmentCountRequired + 1);
+
+    long currentLogSegmentCount = state.index.getLogSegmentCount();
+    writeDataToMeetRequiredSegmentCount(currentLogSegmentCount + extraSegmentCountRequired, null);
+
+    // advance time and delete some data (some of the data will get deleted at deleteTimeMs 
```

Comment:
I don't quite get why do you close the compactor here. This is not the normal flow where in compactor will be closed right? so, why close the compactor here? Same question for above cases too (InterruptionInducingIndex and InterruptionInducingDiskIOScheduler)